src/Razor/src/Microsoft.AspNetCore.Razor.ProjectEngineHost/Utilities/TagHelperCache.cs

```diff
@@ -0,0 +1,109 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.PooledObjects;
+
+namespace Microsoft.AspNetCore.Razor.Utilities;
+
+internal sealed class TagHelperCache
+{
+    public static readonly TagHelperCache Default = new();
+
+    private readonly Dictionary<Checksum, WeakReference<TagHelperDescriptor>> _checksumToTagHelperMap = new();
+
+    private const int CleanUpThreshold = 200;
+    private int _addsSinceLastCleanUp;
+
+    public TagHelperCache()
+    {
+    }
+
+    public TagHelperDescriptor GetOrAdd(Checksum checksum, TagHelperDescriptor tagHelper)
+    {
+        lock (_checksumToTagHelperMap)
+        {
+            if (++_addsSinceLastCleanUp >= CleanUpThreshold)
+            {
+                CleanUpDeadObjects_NoLock();
+            }
+
+            if (!_checksumToTagHelperMap.TryGetValue(checksum, out var weakRef))
+            {
+                _checksumToTagHelperMap.Add(checksum, new(tagHelper));
+                return tagHelper;
+            }
+
+            if (!weakRef.TryGetTarget(out var cachedTagHelper))
+            {
+                weakRef.SetTarget(tagHelper);
+                return tagHelper;
+            }
+
+            return cachedTagHelper;
+        }
+    }
+
+    public bool TryAdd(Checksum checksum, TagHelperDescriptor tagHelper)
+    {
+        lock (_checksumToTagHelperMap)
+        {
+            if (++_addsSinceLastCleanUp >= CleanUpThreshold)
+            {
+                CleanUpDeadObjects_NoLock();
+            }
+
+            if (!_checksumToTagHelperMap.TryGetValue(checksum, out var weakRef))
+            {
+                _checksumToTagHelperMap.Add(checksum, new(tagHelper));
+                return true;
+            }
+
+            if (!weakRef.TryGetTarget(out var cachedTagHelper))
+            {
+                weakRef.SetTarget(tagHelper);
+                return true;
+            }
+
+            return false;
+        }
+    }
+
+    public bool TryGet(Checksum checksum, [NotNullWhen(true)] out TagHelperDescriptor? tagHelper)
+    {
+        lock (_checksumToTagHelperMap)
+        {
+            if (_checksumToTagHelperMap.TryGetValue(checksum, out var weakRef) &&
+                weakRef.TryGetTarget(out tagHelper))
+            {
+                return true;
+            }
+
+            tagHelper = null;
+            return false;
+        }
+    }
+
+    private void CleanUpDeadObjects_NoLock()
+    {
+        using var deadChecksums = new PooledArrayBuilder<Checksum>();
+
+        foreach (var (checksum, weakRef) in _checksumToTagHelperMap)
+        {
+            if (!weakRef.TryGetTarget(out _))
```

Comment:
>weakRef.TryGetTarget(out _) [](http://example.com/codeflow?start=17&length=27)

IsAlive?

Reply:
Actually, it doesn't even look like IsAlive is defined on the generic version of the class. Never mind this suggestion.

Reply:
Actually, after refactoring a bit, the code changed enough that `IsAlive` is no longer appropriate.